### PR TITLE
#HT-122: Invite new users to organization

### PR DESCRIPTION
**Pull request by pranjal5399**

### CODE REVIEW

File: packages/common/source/models/MembershipModel.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            enum: membershipStatuses,
+            required: true,
+        },
+    },
+    {
+        timestamps: true,
+    },
+);
+
+export default model<Membership>("Membership", membershipSchema);
```
Comment (itssamuelrowe) [inline packages/common/source/models/MembershipModel.ts:38]: Perfect!

File: packages/common/source/models/OrganizationModel.ts [hunk 2]
```diff
@@ -5,42 +5,58 @@ import type { Organization } from "../types";
...
+            type: String,
+            minlength: 0,
+            maxlength: 512,
+            default: "",
+        },
+        members: {
+            type: [
+                {
+                    type: Schema.Types.ObjectId,
+                    ref: "User",
```
Comment (itssamuelrowe) [inline packages/common/source/models/OrganizationModel.ts:33]: Let's update this to `Member` instead of `User`.

File: packages/common/source/types/index.ts [hunk 3]
```diff
@@ -295,3 +275,40 @@ export interface Manifest {
     app: App;
     file?: string;
 }
+
+export interface Organization {
+    id: string;
+    name: string;
+    title: string;
+    description: string;
+    members: string[] | User[];
```
Comment (itssamuelrowe) [inline packages/common/source/types/index.ts:284]: `Membership[]` instead of `User[]`.

File: packages/common/source/types/index.ts [hunk 4]
```diff
@@ -295,3 +275,40 @@ export interface Manifest {
...
+    createdAt: Date;
+    updatedAt: Date;
+}
+
+export interface ExternalOrganization {
+    id: string;
+    name: string;
+    title: string;
+    description: string;
+    members: string[] | User[];
```
Comment (itssamuelrowe) [inline packages/common/source/types/index.ts:296]: `Membership[]` instead of `User[]`.

File: packages/common/source/types/index.ts [hunk 5]
```diff
@@ -295,3 +275,40 @@ export interface Manifest {
...
+    updatedAt: Date;
+}
+
+export type OrganizationPage = ExternalListPage<ExternalOrganization>;
+
+export interface Membership {
+    id: string;
+    member: string | User;
+    inviter: string | User;
+    division: string | Group | Organization;
```
Comment (itssamuelrowe) [inline packages/common/source/types/index.ts:309]: Good!

File: packages/common/test/helper/constants.ts [hunk 6]
```diff
@@ -95,4 +95,33 @@ const resources = [
     },
 ];
 
-export { activityLogs, queryTemplates, appData, users, resources };
+const organizations = [
+    {
+        name: "Google",
+        title: "google",
+        description: "",
```
Comment (itssamuelrowe) [inline packages/common/test/helper/constants.ts:102]: Let's not leave the description empty.

File: packages/common/source/types/index.ts [hunk 7]
```diff
@@ -295,3 +275,40 @@ export interface Manifest {
     app: App;
     file?: string;
 }
+
+export interface Organization {
```
Comment (itssamuelrowe) [inline packages/common/source/types/index.ts:279]: You were supposed to derive doc comments based on the information given in the ticket.

File: packages/common/test/models/membership.test.ts [hunk 8]
```diff
@@ -0,0 +1,226 @@
+import { assert } from "chai";
+
+import { MembershipModel } from "../../source/models";
+
```
Comment (itssamuelrowe) [inline packages/common/test/models/membership.test.ts:4]: Remove the line break.

**Review by itssamuelrowe — CHANGES_REQUESTED** [review]

File: packages/api/source/app.ts [hunk 9]
```diff
@@ -1,25 +1,30 @@
...
-    destroy,
-};
\ No newline at end of file
+import express from "express";
+import logger from "morgan";
+import cors from "cors";
+import bodyParser from "body-parser";
+
+import * as graphql from "./graphql";
+
```
Comment (itssamuelrowe) [inline packages/api/source/app.ts:7]: Remove this new line.

File: packages/api/source/app.ts [hunk 10]
```diff
@@ -1,25 +1,30 @@
...
+import { memberships } from "./rest";
+
+const initialize = async () => {
+    const app = express();
+    app.use(cors());
+    app.use(bodyParser.json());
+
+    const router = express.Router();
+    memberships.attachRoutes(router);
```
Comment (itssamuelrowe) [inline packages/api/source/app.ts:16]: You can use `app` instead of creating a new router.

Comment (itssamuelrowe) [reply packages/api/source/app.ts:16]: And move this below `graphql.attachRoutes` call.

File: packages/api/source/controllers/MembershipController.ts [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+import {
+    MembershipModel,
+    UserModel,
+    constants,
+    BadRequestError,
+    NotFoundError,
+    OrganizationModel,
+    ExternalMembership,
+    runAsTransaction,
+    sendEmail,
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:12]: Why do you need `@sendgrid/mail` when you have `sendEmail`?

File: packages/api/source/controllers/MembershipController.ts [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+} from "@hypertool/common";
+import jwt from "jsonwebtoken";
+import sgMail from "@sendgrid/mail";
+
+import { invitationTemplate } from "../utils";
+
+const SENDGRID_API_KEY = process.env.SENDGRID_API_KEY;
+const INVITATION_JWT_SIGNATURE = process.env.INVITATION_JWT_SIGNATURE;
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:20]: Use destructuring:
```
const { ... } = process.env;
```

File: packages/api/source/controllers/MembershipController.ts (hunk 11 above)
Comment (itssamuelrowe) [reply packages/api/source/controllers/MembershipController.ts:12]: Also, `sendEmail` should be in `@hypertool/api`, not `@hypertool/common`. IIRC, I told you to create a `utils` module.

File: packages/api/source/controllers/MembershipController.ts [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+sgMail.setApiKey(SENDGRID_API_KEY);
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:21]: I guess you can move this to `utils`.

File: packages/api/source/controllers/MembershipController.ts [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+            lastName: "<unavailable>",
+            description: "<unavailable>",
+            gender: undefined,
+            countryCode: undefined,
+            emailAddress: emailAddress,
+            emailVerified: true,
+            status: "invited",
+            role: "viewer",
+        });
+
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:76]: Remove the new line.

File: packages/api/source/controllers/MembershipController.ts [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+        await newUser.save();
+        member = newUser;
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:78]: You can directly assign `new UserModel` expression to member. The temporary variable `newUser` is unnecessary.

File: packages/api/source/controllers/MembershipController.ts [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+
+    let membership = await MembershipModel.findOne({
+        member: member.id,
+        organizationId,
+    });
+
+    if (!membership) {
+        const newMembership = new MembershipModel({
+            member: member.id,
+            inviter: "61fcf3cd1ca8c8033509f728",
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:89]: Why is this hard coded?

File: packages/api/source/controllers/MembershipController.ts [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+            division: organizationId,
+            type: "organization",
+            status: "invited",
+        });
+        await newMembership.save();
+        membership = newMembership;
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:95]: Directly assign to `membership`. `newMembership` temporary variable is not necessary.

File: packages/api/source/controllers/MembershipController.ts [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    if (membership.status === "invited") {
+        const token = jwt.sign(
+            { emailAddress, organizationId },
+            INVITATION_JWT_SIGNATURE,
+            {
+                expiresIn: 60 * 60,
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:103]: Add a comment saying how long this is.

File: packages/api/source/controllers/MembershipController.ts [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+            },
+        );
+        const subject = "";
+        const text = "";
+        const html = invitationTemplate(token);
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:108]: You'll have to use string rendering here.

File: packages/api/source/controllers/MembershipController.ts [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+                },
+                {
+                    status: "accepted",
+                },
+            );
+
+            await OrganizationModel.findByIdAndUpdate(
+                organizationId,
+                { $push: { members: user.id } },
+                { safe: true, upsert: true },
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:144]: Why are we upserting? An organization should exist if the user should be invited to it!

File: packages/api/source/rest/index.ts [hunk 21]
```diff
@@ -0,0 +1 @@
+export * as memberships from "./membership.routes";
```
Comment (itssamuelrowe) [inline packages/api/source/rest/index.ts:1]: What's up with `.routes`?

File: packages/api/source/rest/membership.routes.ts [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+const { httpStatuses } = constants;
+
+const attachRoutes = (router: Router): void => {
+    router.get(
+        "/invitation/accept/:jwt",
+        async (request: Request, response: Response, next: NextFunction) => {
+            const { jwt } = request.params;
+            const result = await memberships.verify(null, jwt);
+
+            if (result)
```
Comment (itssamuelrowe) [inline packages/api/source/rest/membership.routes.ts:15]: Where are the braces?

File: packages/api/source/rest/membership.routes.ts [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+                return response
+                    .status(httpStatuses.OK)
+                    .json({ message: "Invitation Accepted." });
```
Comment (itssamuelrowe) [inline packages/api/source/rest/membership.routes.ts:18]: accepted*

File: packages/api/source/utils/index.ts [hunk 24]
```diff
@@ -0,0 +1,228 @@
+const invitationTemplate = (token) => {
```
Comment (itssamuelrowe) [inline packages/api/source/utils/index.ts:1]: Move this to a template file.

Comment (itssamuelrowe) [reply packages/api/source/utils/index.ts:1]: I don't see the point of embedding HTML in JS, when we can separate HTML to a separate file completely.

File: packages/api/source/utils/index.ts [hunk 25]
```diff
@@ -0,0 +1,228 @@
...
+    return `
+<!doctype html>
```
Comment (itssamuelrowe) [inline packages/api/source/utils/index.ts:3]: Send me a screenshot/screen recording of this entire flow.

File: packages/common/source/utils/index.ts [hunk 26]
```diff
@@ -1,5 +1,9 @@
 import mongoose from "mongoose";
+import sgMail from "@sendgrid/mail";
 
+const SENDGRID_API_KEY = process.env.SENDGRID_API_KEY;
+const INVITATION_JWT_SIGNATURE = process.env.INVITATION_JWT_SIGNATURE;
+sgMail.setApiKey(SENDGRID_API_KEY);
```
Comment (itssamuelrowe) [inline packages/common/source/utils/index.ts:6]: Use newlines!

Comment (itssamuelrowe) [reply packages/common/source/utils/index.ts:6]: You need to destructure whenever you write the variable name in two places like this:

const a = object.a;

becomes

const { a } = object;

File: packages/common/source/utils/index.ts [hunk 27]
```diff
@@ -22,6 +26,37 @@ export const extractIds = (items: ObjectWithID[] | string[]): string[] => {
     });
 };
 
+export const sendEmail = async (emailAddress, subject, text, html) => {
+    try {
+        const message = {
+            to: emailAddress,
+            from: { name: "Hypertool", email: "noreply@hypertool.io" },
```
Comment (itssamuelrowe) [inline packages/common/source/utils/index.ts:33]: 1. `sendEmail` should accept an object instead of so many parameters.
2. The from name and email addresses should be specified by the caller. Don't hard code it here.

File: packages/common/source/utils/index.ts [hunk 28]
```diff
@@ -22,6 +26,37 @@ export const extractIds = (items: ObjectWithID[] | string[]): string[] => {
...
+            subject,
+            text,
+            html,
+        };
+
+        await sgMail.send(message);
+    } catch (error) {
+        console.log(error.message);
```
Comment (itssamuelrowe) [inline packages/common/source/utils/index.ts:41]: Why are you catching the error? How will the caller know if there was an error?

File: packages/common/test/helper/constants.ts [hunk 29]
```diff
@@ -95,4 +95,33 @@ const resources = [
     },
 ];
 
-export { activityLogs, queryTemplates, appData, users, resources };
+const organizations = [
+    {
+        name: "Google",
+        title: "google",
+        description: "Let's not leave the description empty.",
```
Comment (itssamuelrowe) [inline packages/common/test/helper/constants.ts:102]: For real?

**Review by itssamuelrowe — CHANGES_REQUESTED** [review]

File: packages/common/test/helper/constants.ts (hunk 29 above)
Comment (pranjal5399) [reply packages/common/test/helper/constants.ts:102]: It's just for testing

File: packages/common/source/utils/index.ts (hunk 27 above)
Comment (pranjal5399) [reply packages/common/source/utils/index.ts:33]: Ok

File: packages/common/source/utils/index.ts (hunk 26 above)
Comment (pranjal5399) [reply packages/common/source/utils/index.ts:6]: Ok

File: packages/api/source/utils/index.ts (hunk 24 above)
Comment (pranjal5399) [reply packages/api/source/utils/index.ts:1]: How will we paas variables then?

File: packages/api/source/rest/membership.routes.ts (hunk 22 above)
Comment (pranjal5399) [reply packages/api/source/rest/membership.routes.ts:15]: added braces. though it works without braces too

File: packages/api/source/controllers/MembershipController.ts [hunk 30]
```diff
@@ -0,0 +1,157 @@
...
+            emailVerified: true,
+            status: "invited",
+            role: "viewer",
+        });
+
+        await member.save();
+    }
+
+    let membership = await MembershipModel.findOne({
+        member: member.id,
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:78]: `member._id`

File: packages/api/source/controllers/MembershipController.ts [hunk 31]
```diff
@@ -0,0 +1,157 @@
...
+        organizationId,
+    });
+
+    if (!membership) {
+        membership = new MembershipModel({
+            member: member.id,
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:84]: `member._id`

File: packages/api/source/controllers/MembershipController.ts [hunk 32]
```diff
@@ -0,0 +1,157 @@
...
+            status: "invited",
+            role: "viewer",
+        });
+
+        await member.save();
+    }
+
+    let membership = await MembershipModel.findOne({
+        member: member.id,
+        organizationId,
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:79]: There is no member called `organizationId` in `MembershipModel`. Has this code been tested?

Comment (itssamuelrowe) [reply packages/api/source/controllers/MembershipController.ts:79]: It should be `division`.

File: packages/api/source/controllers/MembershipController.ts [hunk 33]
```diff
@@ -0,0 +1,157 @@
...
+        const token = jwt.sign(
+            { emailAddress, organizationId },
+            INVITATION_JWT_SIGNATURE,
+            {
+                expiresIn: 60 * 60, // 1 hour
+            },
+        );
+        const subject = "Invitation to join organization";
+        const text = "Text";
+        const html = invitationTemplate({ token });
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:103]: All of this can be directly assigned to the `params` object. Temporary variables are unnecessary.

File: packages/api/source/controllers/MembershipController.ts [hunk 34]
```diff
@@ -0,0 +1,157 @@
...
+        const params = {
+            from: { name: "Hypertool", email: "noreply@hypertool.io" },
+            to: emailAddress,
+            subject,
+            text,
+            html,
+        };
+        sendEmail(params);
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:111]: Shouldn't this call be `await`-ed?

File: packages/api/source/controllers/MembershipController.ts [hunk 35]
```diff
@@ -0,0 +1,157 @@
...
+            member: member.id,
+            inviter: inviterId,
+            division: organizationId,
+            type: "organization",
+            status: "invited",
+        });
+        await membership.save();
+    }
+
+    if (membership.status === "invited") {
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:93]: I think you were supposed to throw an exception when the status is anything other than `invited`. Please refer the ticket.

File: packages/api/source/controllers/MembershipController.ts [hunk 36]
```diff
@@ -0,0 +1,157 @@
...
+                { emailAddress },
+                {
+                    $set: { organization: organizationId },
+                },
+                { new: true },
+            );
+
+            await MembershipModel.findOneAndUpdate(
+                {
+                    member: user.id,
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:135]: `user._id`

Let's use `object._id` everywhere because there will be consistency with both lean and non-lean objects. Only client-side code will use `object.id`, which is why we convert `_id` to `id` in `toExternal`.

File: packages/api/source/controllers/MembershipController.ts [hunk 37]
```diff
@@ -0,0 +1,157 @@
...
+                    organizationId,
+                },
+                {
+                    status: "accepted",
+                },
+            );
+
+            await OrganizationModel.findByIdAndUpdate(
+                organizationId,
+                { $push: { members: user.id } },
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:145]: `user._id`

File: packages/api/source/controllers/MembershipController.ts [hunk 38]
```diff
@@ -0,0 +1,157 @@
...
+                { safe: true },
+            );
+        });
+
+        return true;
+    } catch (error) {
+        //throw new BadRequestError(error.message);
+        return false;
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:153]: I think I already left a comment here. :/

File: packages/api/source/graphql/privateAPI.ts [hunk 39]
```diff
@@ -272,7 +275,7 @@ const typeDefs = gql`
 
     type QueryTemplatePage {
         totalRecords: Int!
-        totalPages: Int!
+        totalPages: Int!
```
Comment (itssamuelrowe) [inline packages/api/source/graphql/privateAPI.ts:278]: Why did you add space here?

File: packages/api/source/graphql/privateAPI.ts [hunk 40]
```diff
@@ -312,6 +315,25 @@ const typeDefs = gql`
         error: GraphQLJSON,
     }
 
+    enum MembershipTypes{
```
Comment (itssamuelrowe) [inline packages/api/source/graphql/privateAPI.ts:318]: `enum MembershipTypes {`

Where's the space before the brace? :/

File: packages/api/source/graphql/privateAPI.ts [hunk 41]
```diff
@@ -312,6 +315,25 @@ const typeDefs = gql`
...
+        ${membershipTypes.join("\n")}
+    }
+
+    enum MembershipStatuses{
```
Comment (itssamuelrowe) [inline packages/api/source/graphql/privateAPI.ts:322]: Missing space!

File: packages/api/source/graphql/privateAPI.ts [hunk 42]
```diff
@@ -312,6 +315,25 @@ const typeDefs = gql`
...
+        ${membershipStatuses.join("\n")}
+    }
+
+    type Membership {
+        id: String!
```
Comment (itssamuelrowe) [inline packages/api/source/graphql/privateAPI.ts:327]: IDs are supposed to be of type `ID` not `String`.

File: packages/api/source/graphql/privateAPI.ts [hunk 43]
```diff
@@ -312,6 +315,25 @@ const typeDefs = gql`
...
+        member: String!
```
Comment (itssamuelrowe) [inline packages/api/source/graphql/privateAPI.ts:328]: You need to use proper types! Member is of type `User`, not `String`. Make sure all the other types in this schema type follow this.

File: packages/api/source/graphql/privateAPI.ts [hunk 44]
```diff
@@ -312,6 +315,25 @@ const typeDefs = gql`
...
+        inviter: String!
+        division: String!
```
Comment (itssamuelrowe) [inline packages/api/source/graphql/privateAPI.ts:330]: The types are incorrect.

File: packages/api/source/graphql/privateAPI.ts [hunk 45]
```diff
@@ -443,6 +465,12 @@ const typeDefs = gql`
             context: GraphQLJSON
             component: ComponentOrigin!
         ): ActivityLog!    
+
+        createMembership(
+            emailAddress: String,
```
Comment (itssamuelrowe) [inline packages/api/source/graphql/privateAPI.ts:470]: `emailAddress` is mandatory. So why haven't you added `!`?

File: packages/api/source/rest/membership-routes.ts [hunk 46]
```diff
@@ -0,0 +1,28 @@
...
+import { memberships } from "../controllers";
+
+const { httpStatuses } = constants;
+
+const attachRoutes = (router: Router): void => {
+    router.get(
+        "/invitation/accept/:jwt",
+        async (request: Request, response: Response, next: NextFunction) => {
+            const { jwt } = request.params;
+            const result = await memberships.verify(null, jwt);
```
Comment (itssamuelrowe) [inline packages/api/source/rest/membership-routes.ts:13]: What's the `null` for?

File: packages/api/source/rest/membership-routes.ts [hunk 47]
```diff
@@ -0,0 +1,28 @@
...
+
+            if (result) {
+                return response
+                    .status(httpStatuses.OK)
+                    .json({ message: "Invitation acepted." });
```
Comment (itssamuelrowe) [inline packages/api/source/rest/membership-routes.ts:18]: Typo

File: packages/api/source/utils/index.ts [hunk 48]
```diff
@@ -0,0 +1,28 @@
...
+        const message = {
+            to,
+            from,
+            subject,
+            text,
+            html,
+        };
+
+        await sgMail.send(message);
+        //console.log(res);
```
Comment (itssamuelrowe) [inline packages/api/source/utils/index.ts:21]: Don't include log statements in the commit history. Make sure you see the diff before making commits.

File: packages/api/source/utils/index.ts [hunk 49]
```diff
@@ -0,0 +1,28 @@
...
+    } catch (error) {
+        console.log(error.message);
```
Comment (itssamuelrowe) [inline packages/api/source/utils/index.ts:23]: Again, don't include logs in the commit history.

File: packages/api/source/utils/templates.ts [hunk 50]
```diff
@@ -0,0 +1,228 @@
+import Mustache from "mustache";
+
+export const invitationTemplate = (data) => {
+    const { token } = data;
+    return Mustache.render(
+        `<!doctype html>
+    <html>
+
```
Comment (itssamuelrowe) [inline packages/api/source/utils/templates.ts:8]: How many times should I tell? This needs to be in a template file that's going to be loaded using `fs`.

Comment (itssamuelrowe) [reply packages/api/source/utils/templates.ts:8]: If you're going to use string interpolation then why is Mustache necessary? :/

Comment (itssamuelrowe) [reply packages/api/source/utils/templates.ts:8]: Read the documentation for Mustache, man.

File: packages/common/source/models/MembershipModel.ts [hunk 51]
```diff
@@ -0,0 +1,43 @@
+import { Schema, model } from "mongoose";
+
+import { Membership } from "../types";
+import { membershipStatuses, membershipTypes } from "../utils/constants";
+
+const membershipSchema = new Schema(
+    {
+        // An identifier that points to the User whose membership is being defined by the current document.
```
Comment (itssamuelrowe) [inline packages/common/source/models/MembershipModel.ts:8]: How many times should I tell you to not use `//` comments? Replace all these comments with `/* ... */`. Also, document the interfaces. Use doc comments there. Learn about doc comments before you continue.

File: packages/common/source/models/MembershipModel.ts [hunk 52]
```diff
@@ -0,0 +1,43 @@
...
+            ref: "User",
+            required: true,
+        },
+        // An identifier that points to the User that invited the member to the class specified by division.
+        inviter: {
+            type: Schema.Types.ObjectId,
+            ref: "User",
+            required: true,
+        },
+        // An identifier that points to the division. This attribute is polymorphic, that is, its meaning is defined based on the value of type attribute. For example, if type is organization, then the identifier points to an organization document. On the other hand, if type is group, then the identifier points to a group document.
```
Comment (itssamuelrowe) [inline packages/common/source/models/MembershipModel.ts:20]: Don't write lines longer than 100 characters. IIRC, our workspace is configured to render a margin in VS Code for 100 characters.

File: packages/common/source/models/OrganizationModel.ts [hunk 53]
```diff
@@ -5,42 +5,64 @@ import type { Organization } from "../types";
...
-      maxlength: 512,
-      default: "",
-    },
-    users: {
-      type: [
-        {
-          type: Schema.Types.ObjectId,
-          ref: "User",
+    {
+        // An identifier that helps humans identify the organization across Hypertool.
```
Comment (itssamuelrowe) [inline packages/common/source/models/OrganizationModel.ts:9]: We don't prefer single-line comments.

**Review by itssamuelrowe — CHANGES_REQUESTED** [review]

File: packages/common/source/models/OrganizationModel.ts (hunk 53 above)
Comment (pranjal5399) [reply packages/common/source/models/OrganizationModel.ts:9]: This is taken from the ticket

File: packages/api/source/graphql/privateAPI.ts (hunk 39 above)
Comment (pranjal5399) [reply packages/api/source/graphql/privateAPI.ts:278]: by mistake i gues

File: packages/api/source/controllers/MembershipController.ts [hunk 54]
```diff
@@ -0,0 +1,161 @@
...
+        await member.save();
+    }
+
+    let membership = await MembershipModel.findOne({
+        member: member._id,
+        division: organizationId,
+    });
+
+    if (membership) {
+        await invite(emailAddress, organizationId);
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:83]: Remove this.

File: packages/api/source/controllers/MembershipController.ts [hunk 55]
```diff
@@ -0,0 +1,161 @@
...
+        throw new BadRequestError(
+            "User is trying to create duplicate invitation.",
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:85]: "Cannot create a duplicate invitation."

File: packages/api/source/controllers/MembershipController.ts [hunk 56]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    membership = new MembershipModel({
+        member: member._id,
+        inviter: inviterId,
+        division: organizationId,
+        type: "organization",
+        status: "invited",
+    });
+
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:96]: Remove this line.

File: packages/api/source/controllers/MembershipController.ts [hunk 57]
```diff
@@ -0,0 +1,161 @@
...
+    );
+    const params = {
+        from: { name: "Hypertool", email: "noreply@hypertool.io" },
+        to: emailAddress,
+        subject: "Invitation to join organization",
+        text: "Text",
+        html: await invitationTemplate({ token }),
+    };
+    await sendEmail(params);
+};
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:120]: Remove this. Let's not have reinvite right now.

File: packages/api/source/controllers/MembershipController.ts [hunk 58]
```diff
@@ -0,0 +1,161 @@
...
+
+const verify = async (context, token): Promise<Boolean> => {
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/MembershipController.ts:122]: Add type annotations.

File: packages/api/source/controllers/OrganizationControllers.ts [hunk 59]
```diff
@@ -33,18 +33,30 @@ const filterSchema = joi.object({
...
+        apps,
+    } = organization;
 
     return {
         id: id || _id.toString(),
         name,
         description,
-        users: extractIds(users),
+        title,
+        apps,
```
Comment (itssamuelrowe) [inline packages/api/source/controllers/OrganizationControllers.ts:58]: Call `extractIds` on `apps`.

File: packages/api/source/graphql/privateAPI.ts [hunk 60]
```diff
@@ -324,6 +327,25 @@ const typeDefs = gql`
...
+    }
+
+    enum MembershipStatuses {
+        ${membershipStatuses.join("\n")}
+    }
+
+    type Membership {
+        id: ID!
+        member: ID!
+        inviter: ID!
```
Comment (itssamuelrowe) [inline packages/api/source/graphql/privateAPI.ts:341]: Add relevant types here.

File: packages/api/source/rest/membership-routes.ts [hunk 61]
```diff
@@ -0,0 +1,32 @@
+import type { Router, Request, Response, NextFunction } from "express";
+import { constants } from "@hypertool/common";
```
Comment (itssamuelrowe) [inline packages/api/source/rest/membership-routes.ts:2]: Newline.

File: packages/api/source/utils/index.ts [hunk 62]
```diff
@@ -0,0 +1,27 @@
...
+            to,
+            from,
+            subject,
+            text,
+            html,
+        };
+
+        await sgMail.send(message);
+    } catch (error) {
+        throw new BadRequestError(error.message);
```
Comment (itssamuelrowe) [inline packages/api/source/utils/index.ts:23]: Change to `Error`.

File: packages/api/source/utils/templates.ts [hunk 63]
```diff
@@ -0,0 +1,10 @@
+import Mustache from "mustache";
+import fs from "fs";
+
+const fsPromisify = fs.promises;
+const filePath = `${process.cwd()}/source/templates/invitation-template.html`;
```
Comment (itssamuelrowe) [inline packages/api/source/utils/templates.ts:5]: Rename the variable to "invitationTemplatePath".

File: packages/api/source/utils/templates.ts [hunk 64]
```diff
@@ -0,0 +1,10 @@
...
+
+export const invitationTemplate = async (data) => {
+    const str = await fsPromisify.readFile(filePath, "utf8");
```
Comment (itssamuelrowe) [inline packages/api/source/utils/templates.ts:8]: No abbreviations. `str`

File: packages/common/source/models/MembershipModel.ts [hunk 65]
```diff
@@ -0,0 +1,51 @@
+import { Schema, model } from "mongoose";
+
+import { Membership } from "../types";
```
Comment (itssamuelrowe) [inline packages/common/source/models/MembershipModel.ts:3]: Add `type` to the type import.

File: packages/common/source/models/MembershipModel.ts [hunk 66]
```diff
@@ -0,0 +1,51 @@
...
+        /* The status of the membership. Valid values are as follows: accepted, deleted, banned, and invited.
+         */
+        status: {
+            type: String,
+            enum: membershipStatuses,
+            required: true,
+        },
+    },
+    {
+        timestamps: true,
```
Comment (itssamuelrowe) [inline packages/common/source/models/MembershipModel.ts:47]: Add `immutable` to `createdAt`.

File: packages/common/source/models/MembershipModel.ts [hunk 67]
```diff
@@ -0,0 +1,51 @@
...
+        inviter: {
+            type: Schema.Types.ObjectId,
+            ref: "User",
+            required: true,
+        },
+        /* An identifier that points to the division.
+         * This attribute is polymorphic, that is, its meaning is defined based on the value of type attribute.
+         * For example, if type is organization, then the identifier points to an organization document.
+         * On the other hand, if type is group, then the identifier points to a group document.
+         */
```
Comment (itssamuelrowe) [inline packages/common/source/models/MembershipModel.ts:26]: 80 characters per line.

File: packages/common/source/models/OrganizationModel.ts [hunk 68]
```diff
@@ -5,42 +5,70 @@ import type { Organization } from "../types";
...
         },
-      ],
-      required: true,
     },
-    status: {
-      type: String,
-      enum: organizationStatuses,
-      default: "active",
+    {
+        timestamps: true,
```
Comment (itssamuelrowe) [inline packages/common/source/models/OrganizationModel.ts:66]: Add `immutable` to `createdAt`.

File: packages/common/source/models/OrganizationModel.ts [hunk 69]
```diff
@@ -5,42 +5,70 @@ import type { Organization } from "../types";
...
+        /*An identifier that helps humans identify the organization across Hypertool.
+         */
+        name: {
+            type: String,
+            minlength: 1,
+            maxlength: 256,
+            required: true,
+            trim: true,
+        },
+        /*The display name of the organization.
```
Comment (itssamuelrowe) [inline packages/common/source/models/OrganizationModel.ts:18]: Space after `/*`.

**Review by itssamuelrowe — CHANGES_REQUESTED** [review]

**Review by itssamuelrowe — APPROVED** [review]
